src/ui/ui-kit/Button/HoldableButton.tsx

```diff
@@ -0,0 +1,229 @@
+    onClick,
+    holdColor,
+    successColor,
+    errorColor,
+    text,
+    submittingText = 'Submitting...',
+    successText = 'Success',
+    errorText = 'Something was wrong',
+    holdDuration = 1000,
+    submitting,
+    success,
+    error,
+    holdHint = 'Please press and hold the button',
+    ...props
+  }: Props & { as?: As } & ComponentPropsWithoutRef<As> & {
+      ref?: ComponentPropsWithRef<As>['ref'];
+    },
+  ref: React.Ref<ComponentPropsWithRef<As>['ref']>
+) => {
+  const realButtonRef = useRef<HTMLButtonElement | null>(null);
+  const [innerState, setInnerState] = useState<'idle' | 'hold'>('idle');
+  const [showHint, setShowHint] = useState(false);
+
+  const state: State = submitting
+    ? 'submitting'
+    : success
+    ? 'success'
+    : error
+    ? 'error'
+    : innerState;
+  const holdTimerRef = useRef<number | ReturnType<typeof setTimeout>>(0);
+  const holdDurationCounter = useRef<number>(0);
+  const lastDisplayHintTime = useRef<number>(0);
+  const shortHoldCounter = useRef<number>(0);
+
+  const displayHint = useCallback(() => {
+    lastDisplayHintTime.current = Date.now();
+    shortHoldCounter.current = 0;
+    setShowHint(true);
+    setTimeout(() => setShowHint(false), HINT_SHOW_DURATION);
+  }, []);
+
+  const handleMouseDown = useCallback(() => {
+    setInnerState('hold');
+    holdDurationCounter.current = Date.now();
+    holdTimerRef.current = setTimeout(() => {
+      realButtonRef.current?.click();
+      setInnerState('idle');
+    }, holdDuration);
+  }, [holdDuration]);
+
+  const handleMouseUp = useCallback(() => {
+    if (Date.now() - holdDurationCounter.current < SHORT_PRESS_DURATION) {
+      shortHoldCounter.current += 1;
+    }
+    if (
+      shortHoldCounter.current > 2 &&
+      Date.now() - lastDisplayHintTime.current > HINT_SHOW_MIN_BREAK
+    ) {
+      displayHint();
+    }
+
+    clearTimeout(holdTimerRef.current);
+    setInnerState('idle');
+  }, [displayHint]);
+
+  const isButton = as == null || as === 'button';
+  return (
+    <>
+      <UIText
+        as={as || 'button'}
+        type="button"
+        ref={ref}
+        kind="small/accent"
+        className={cn(
+          className,
+          buttonStyles[kind],
+          buttonStyles.button,
+          styles.button,
+          {
+            [buttonStyles.asButton]: !isButton,
+            [styles.hold]: state === 'hold',
+            [styles.submitting]: state === 'submitting',
+            [styles.success]: state === 'success',
+            [styles.error]: state === 'error',
+          }
+        )}
+        onMouseDown={(e) => {
+          if (e.button === 0) {
+            e.preventDefault();
+            handleMouseDown();
+          }
+        }}
+        onMouseUp={handleMouseUp}
+        onKeyDown={(e) => {
+          e.preventDefault();
+          if (e.key === 'Enter' && innerState === 'idle') {
+            handleMouseDown();
+          }
+        }}
+        onKeyUp={handleMouseUp}
```

Comment:
1. (harmless) If I press any keyboard button and hold it, then hold click, then release keyboard button, the action will cancel
2. (dangerous, probably a bug) If I start holding "Enter", then do a quick click (click and release), then release enter, the animation seemingly cancels, but then fills up and completes! And transaction submits

Reply:
Hm, this is interesting. I didn't want to overcomplicate this component. The second one is a bug, however, I am not sure that it is really a dangerous. We tried to prevent misclicks, not beat the case when user is playing with this button and challenging is. I can think about a fix here or we can postpone it for later.

Reply:
The point of this button is to prevent accidental submissions. If it can create accidental submission, we should fix that